### ISSUE TITLE
Do not start pm in upgrade mode

### ISSUE THREAD

**Issue by CODeRUS** [sailfishos-patches/patchmanager#39]
Add os-update.target to conflicts of pm service

**Comment by Olf0** [comment]
Thanks, this potentially closes issue #33, too.
Will test.

**Comment by nephros** [comment]
I'm sorry, I'd like to get some context on this.

Change proposed is to add a `Conflicts=os-update.target ` entry to `/usr/lib/systemd/system/dbus-org.SfietKonstantin.patchmanager.service`, correct?

With the goal to not interfere with anything going on during a system update.

Qs:

1. is this still relevant in the days of PM3 and SFOS 4.x?
2. if yes, what creates os-update.target? Is it reliably* a thing that indicates a device is currently updating?
2. if not, what else might be used?

[*] with definitions of reliable adapted to the reality of how SFOS/Jolla changes happen... ;)

**Comment by Olf0** [comment]
As:
1. Yes, I believe so: It does not make any sense for PM's backgroud services to start, when `os-update.target` is active.
2. a. SailfishOS!
    b. Well, it is reliably indicating that a SailfishOS installation has booted into the "single user upgrade phase" of an OS upgrade via GUI.
        You know (if not upgrading SFOS at the command line), one looks for an SFOS-upgrade in the Settings app, downloads its RPMs there; then the device reboots to `os-update.target`, only a progress bar is shown while the RPMs are installed and finally the device switches (or reboots, I do not remember precisely) to the regular multiuser-with-UI target.
       Mind that this method of upgrading SFOS cannot be used on community ports and "Sailfish Free / Trial" installations (i.e., SFOS installations without a paid license from Jolla) and there are may other ways of upgrading SFOS (the "zypper dance" / per `zypper --dup`, per `version dup`, per `pkcon upgrade-system`, which all not boot to a special target.
     TL;DR: `os-update.target` is a kind of "single user mode for upgrading SFOS", used when triggering an OS upgrade per Settings app.
  c. Well, this mechanism exists at least since SFOS 1.1.x (when I joined the club). 
3. Good question.  I guess the answer is "nothing", because the upgrading per command line (whichever way) cannot be detected, as it is just downloading and installing RPMs.
   Actually, what all "alternative" methods do have in common, is that one has to set the SSU utility into release mode per `ssu re <version-to-upgrade-to>` while at the regular multiuser-with-UI target (the GUI upgrade likely does also set SSU to release mode, but only while at the `os-update.target`).  The Seamless Software Update (SSU) utility was [originally developed by Nokia for the N900](https://wiki.maemo.org/Seamless_Software_Update) and [Jolla's documentation for it is extremely terse](https://sailfishos.org/wiki/Seamless_Software_Update_Cellular_Usage), but their [MDM documentation](https://sailfishos.org/develop/docs/sailfish-mdm/ssu.html/) that there is an "*int deviceMode() - The repository configuration mode e.g. Release, RnD, Upgrade. See ssu docs for further info.*".  IIRC SSU is proprietary (not 100% sure).

**Comment by nephros** [comment]
Thank you. Much clearer now.

Can you please test https://github.com/nephros/patchmanager/commit/d14a3ee9287fc48f48668d0978be92ee7edaef10 somehow? (Well, it's just the single line added :) ).
I have confirmed adding the line doesn't affect restarting the service or rebooting, but I do not know how to test the actual "os-update.target exists" case.

**Comment by Olf0** [comment]
> Thank you. Much clearer now.

Well, then it is time for me to report the things I do not understand WRT @CODeRUS' original issue description, "[*Add os-update.target to conflicts of pm service*](https://github.com/sailfishos-patches/patchmanager/issues/39#issue-422693087)": See below.

> @Olf0 Can you please test [nephros@d14a3ee](https://github.com/nephros/patchmanager/commit/d14a3ee9287fc48f48668d0978be92ee7edaef10) somehow? (Well, it's just the single line added :) ). I have confirmed adding the line doesn't affect restarting the service or rebooting, but I do not know how to test the actual "os-update.target exists" case.

Just adding `Conflicts=os-update.target` (to `bin/patchmanager-daemon/systemd/dbus-org.SfietKonstantin.patchmanager.service`) is definitely too simple:
1. As with *all* dependency statements in systemd units, `Conflicts=` does not include an order and does not make much sense (or results in unexpected behaviour) without one.  See https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Conflicts=
   Hence an `After=os-update.target` or `Before=os-update.target` should be added, but which? → See below.
2. I tried research how this is supposed to work in practice (aside of the [theoretical considerations](https://github.com/sailfishos-patches/patchmanager/issues/39#issuecomment-930677139) I described above).
   * Searching the WWW for `"os-update.target"` reveals ... nothing.
   * But I remembered that systemd has some mechanisms Jolla might utilise:
     * The "systemd.special" [`system-update.target`](https://www.freedesktop.org/software/systemd/man/systemd.special.html#system-update.target)
     * ... which is part of [systemd's mechanisms to handle "offline updates"](https://www.freedesktop.org/software/systemd/man/systemd.offline-updates.html).
       I *assumed* that Jolla utilises that, hence aforementioned "theoretical considerations" were based on what I remembered of this man page.
    * So I tried to look at this in practice on an XperiaX@SFOS3.2.1
      But executing a `systemctl -l -t target --all` (or an equivalent `systemctl -l -t target --all list-units`) does not show an `os-update.target`!
      Neither does a `systemctl -l -t target --all list-unit-files` show anything indicating such a target!

But at least re-reading [systemd's man-page for "offline updates"](https://www.freedesktop.org/software/systemd/man/systemd.offline-updates.html), specifically "Recommendation 4" convinces me that `Before=os-update.target` is the better ordering statement.
Unfortunately `system-update.target` is completely missing at systemd's man-page "[bootup](https://www.freedesktop.org/software/systemd/man/bootup.html)".

@CODeRUS, can you please provide some background on `os-update.target`, how it works, where it can be found and analysed, and if simply adding `Conflicts=os-update.target` and `Before=os-update.target` to `bin/patchmanager-daemon/systemd/dbus-org.SfietKonstantin.patchmanager.service` is what you had in mind.

**Comment by Olf0** [comment]
> One more thing:
> 
> Our RPM installs or changes `/etc/ld.so.preload`, which will of course be in effect always, even in upgrade mode.

But that behaves transparently (i.e., like its original version), when it loads things not on a whitelist, correct?
Or does nothing call it, when the PM-unit is not active?

Sorry, I never looked those things up and have not yet thought about them.

"installs *or* changes" (i.e., which of both) is the first thing we should know for sure.

> We _could_ introduce a special service which `Requires/Before=system-update-pre.target` which moves that file out of the way during updates, so failure to load the library doesn't kill the update process.

Which "the library"?
(Will ponder about that, some time.)

> Or we could set NO_PM_PRELOAD as an environment variable, but I don't know how to set that globally for everything and only in system-update mode.

I do not think that is viable.
